src/helpers.js

```diff
@@ -1,85 +1,21 @@
-const globalObj = typeof window === 'undefined' ? global : window
 // Constant node.nodeType for text nodes, see:
 // https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType#Node_type_constants
 const TEXT_NODE = 3
 
-// Currently this fn only supports jest timers, but it could support other test runners in the future.
-function runWithRealTimers(callback) {
-  return hasJestTimers()
-    ? runWithJestRealTimers(callback).callbackReturnValue
-    : // istanbul ignore next
-      callback()
-}
-
-function hasJestTimers() {
-  return (
-    typeof jest !== 'undefined' &&
-    jest !== null &&
-    typeof jest.useRealTimers === 'function'
-  )
-}
-
-function runWithJestRealTimers(callback) {
-  const timerAPI = {
-    clearInterval,
-    clearTimeout,
-    setInterval,
-    setTimeout,
-  }
-
-  // For more on why we have the check here,
-  // checkout https://github.com/testing-library/dom-testing-library/issues/914
-  if (typeof setImmediate === 'function') {
-    timerAPI.setImmediate = setImmediate
-  }
-  if (typeof clearImmediate === 'function') {
-    timerAPI.clearImmediate = clearImmediate
-  }
-
-  jest.useRealTimers()
-
-  const callbackReturnValue = callback()
-
-  const usedFakeTimers = Object.entries(timerAPI).some(
-    ([name, func]) => func !== globalObj[name],
-  )
-
-  if (usedFakeTimers) {
-    jest.useFakeTimers(timerAPI.setTimeout?.clock ? 'modern' : 'legacy')
-  }
-
-  return {
-    callbackReturnValue,
-    usedFakeTimers,
-  }
-}
-
 function jestFakeTimersAreEnabled() {
-  return hasJestTimers()
-    ? runWithJestRealTimers(() => {}).usedFakeTimers
-    : // istanbul ignore next
-      false
-}
-
-// we only run our tests in node, and setImmediate is supported in node.
-// istanbul ignore next
-function setImmediatePolyfill(fn) {
-  return globalObj.setTimeout(fn, 0)
-}
-
-function getTimeFunctions() {
-  // istanbul ignore next
-  return {
-    clearTimeoutFn: globalObj.clearTimeout,
-    setImmediateFn: globalObj.setImmediate || setImmediatePolyfill,
-    setTimeoutFn: globalObj.setTimeout,
+  /* istanbul ignore else */
+  if (typeof jest !== 'undefined' && jest !== null) {
+    return (
+      // legacy timers
+      setTimeout._isMockFunction === true ||
+      // modern timers
+      Object.prototype.hasOwnProperty.call(setTimeout, 'clock')
```

Comment:
Neither check guarantees that the timer is not mocked by other means.
We've already been there and it caused some confusion:
https://github.com/testing-library/dom-testing-library/blob/bc65b91fed81d900e8eb2aff1813752a4f487869/src/helpers.js#L22-L51

Reply:
What other means would that be and why should we support them?

Reply:
Additionally to what @eps1lon said, as long as the name of this function is `**jest**FakeTimersAreEnabled`, I think the check covers all possibilities.  If someone mocks timers through another library, or manually, it seems beyond the scope of this method (feels like it needs to be a separate feature).

Reply:
I agree with @9still here.

Reply:
My concern is that this function might return `true` although there are no jest fake timers in place.
That's what the previous solution fixed.

If anyone later relies on this function to tell if jest timers are active and acts on this cognition, we go full circle.
The side-effect described in #884 happened due to `runWithRealTimers` relying on `getJestFakeTimersType` to do what its name said.

So if we can't get rid of a function called `jestFakeTimersAreEnabled`, we should ensure that it only returns `true` if jest fake timers are enabled.

Reply:
Sorry, I'm not sure I understand how this could return `true` without fake timers in place 🤔 It looks like this only would return true in a jest environment if `setTimeout._isMockFunction` or `setTimeout.clock` is defined. I suppose someone could mock `setTimeout` themselves (that's odd) or add a `.clock` to `setTimeout` (also odd). Are you suggesting that this should basically be: `return Boolean(getJestFakeTimersType())`?

Reply:
That's what happened before. `jest.useFakeTimers('modern')` was used in another test and then this helper recognized the fake timer to be enabled. `runWithRealTimers` then acted upon this and caused a leak.

Given `runWithRealTimers` is removed the problem might be less likely to resurface again.
But who knows what might be implemented in the future? I would not want to keep in mind that the `jestFakeTimersAreEnabled` helper is not reliable.

Better remove the helper completely or keep the implementation that is safe to use no matter what the user or any other library did to `global.setTimeout`.

Reply:
> Better remove the helper completely or keep the implementation that is safe to use no matter what the user or any other library did to global.setTimeout.

I think in the future it might be better to explicitly configure the clock and then just try to detect the clock when using jest in its default configuration. But we should probably do this in a follow-up. 

Right now I just hope we can land this without too much breakage. Especially since we can remove so much code while also closing 3 issues.

Reply:
> I don't understand why we would risk reintroducing a bug just for removing the lines that have been explicitly introduced to prevent this.

Let's unpack this a bit. I'm not explicitly risking a bug by removing the code. I just think the previous code didn't serve any purpose. This assumption holds with our test suite. Hence why I'm asking what cases I am missing.

I do risk a bug implicitly though just by virtue of changing code. I'm very well aware of that. Considering this changes gets rid of a whopping 300 lines of code while fixing 3 open issues, I'm fairly confident the risk is worth it. Any risk remaining will be tested in the alpha phase.

> Could we not just update the implementation if there are issues with jest27 ?

There are issues with jest 27. No "if". Obviously I would prefer to "just update the implementation". But as far as I can tell the previous  implementation is incompatible with jest 27 while also causing existing issues (https://github.com/testing-library/dom-testing-library/issues/880, https://github.com/testing-library/dom-testing-library/issues/756). The proposed implementation in this PR is compatible with jest 26 and jest 27.

Reply:
I haven't spent enough time examining this to tell if this will actually cause problems or not with the current `waitFor` implementation.

I just stumbled over that revert to the old check and remembered how difficult it was to identify the culprit for the leaking test last time because the code seemed too simple and it needed a specific setup to fail.
I'm not confident that an alpha phase will catch similar things - last time the insufficient check was introduced Sep, 2019 and extended to modern versions (causing the bug) in Jun, 2020 - the issue was tracked down in Feb, 2021 just because the bug was also relevant in this library's own test suite. But it would not throw an error - just subsequent tests were silently ignored. (If it wasn't for test coverage, that might have went unnoticed.)

If we can kick out the remainder of the helper, I'm all for it.
But I would hate to require everybody to remember that this helper should not be used for the thing its name suggests.

> There are issues with jest 27. No "if". Obviously I would prefer to "just update the implementation". But as far as I can tell the previous implementation is incompatible with jest 27 while also causing existing issues (#880, #756). The proposed implementation in this PR is compatible with jest 26 and jest 27.

I haven't seen yet which part of the helper is incompatible with jest27. Could you give me a hint which part is considered incompatible? I'll look into it later and try to come up with a fix.
(I checked out the branch yesterday. When updating the `kcd-scripts` tests would fail everywhere. How did you run the test suite with jest27?)

Reply:
> I haven't spent enough time examining this to tell if this will actually cause problems or not with the current waitFor implementation.

Would you agree that it may be more constructive to find these issues first, submit a test and then refer to it here instead of voicing concern first and putting the burden on me?

> I haven't seen yet which part of the helper is incompatible with jest27.

Every part.

> How did you run the test suite with jest27?

See https://github.com/testing-library/dom-testing-library/pull/962 as explained in the initial issue description

Reply:
> Would you agree that it may be more constructive to find these issues first, submit a test and then refer to it here instead of voicing concern first and putting the burden on me?

I did find the issue, submitted a test that covered that scenario and you deleted it with this change.
And then I raised the concern that deleting that test and reverting to a code snippet that failed before might fail in the future.

I assumed you know best about the part of the code you edited.
I just know from time previously spent on this that this old helper implementation isn't reliable.
It doesn't exactly do what it suggests. That's why I think it shouldn't be in the codebase.

> > I haven't seen yet which part of the helper is incompatible with jest27.
> 
> Every part.

So much for constructive... :facepalm: 
According to the docs jest-next fake timers work like before - I'll check their code to see if every part changed.

Reply:
> I did find the issue, submitted a test that covered that scenario and you deleted it with this change.

Could you re-submit it as an integration test? I just deleted the unit test that become obsolete with the removal of `useRealTimers`. As far as I can tell this function isn't part of the public API and therefore removing tests is safe if the function is removed.

> I assumed you know best about the part of the code you edited.

Yeah, you shouldn't do that for any edit I make. I don't pressume to know the code I edit to know "best". Otherwise I wouldn't ask for review or collaborate with other people in general.

> So much for constructive... facepalm

May I remind you to be respectul. I don't know how else to say this. The tests were just failing left and right. Maybe there is one particular line in jest that changed but I shouldn't need to debug their implementation because their implementation is irrelevant to our usage. If you want to work on another implementation then please go ahead. In the end I don't care about the implementation. Only about the observable behavior. But I'm not doing that work for you just because you don't like the current implementation.

I know this is kind of easy to say for me but don't take removal of your code personally. Stuff just changes over time. Just because we remove code, doesn't mean the code was useless. It obviously served a purpose in the past. But it isn't unlikely that implementation details change over time.

Reply:
Keep twisting words. Your standards for respectful behavior seem to differ quite a bit depending on if your addressor or addressee.

You might consider this a fundamentally incorrect assumption, but I'll continue to assume that someone is competent and has thought well about a change before writing it. So unless I'm certain to know better I will not easily disregard their thoughts and rather raise a question than redo their work to prove them wrong.

If you consider questions and acknowledgement of limitations of ones one knowledge as unconstructive and a burden to you, that might constrain the possible collaboration quite a bit.

Now you can go ahead and just delete the code because the only possible reason one might be concerned about the removal is the code being precious and taking the removal personally.

Reply:
> I haven't seen yet which part of the helper is incompatible with jest27.

@ph-fritsche From the point of view of a user - if you are referring to the implementation added in #887, it completely broke modern timers (& they became the default in Jest 27).   #961 describes the reason in detail, but basically half of our fake timer tests would hang with this implementation.

Reply:
@9still Thanks a lot for that detailed report.

That's one ugly mess with those timers.
The short and simple suggested solution is not feasible. Its the root of [this](https://github.com/ph-fritsche/dom-testing-library/blob/24ebf282d10782b3d0a22b939a1b2950d4b5bd1b/src/__tests__/helpers.js#L86-L118).

`jest.useRealTimers` is the only reliable way to tell if jest's fake timers are enabled - at least the only one I could find. Preserving that `clock` as you suggested might work.

I like the option to plug in your own callback for advancing timers during `waitFor`. It's the cleanest solution, but I guess users would like a sensible fallback.

Maybe the "guess" per the simple solution is enough for the new `waitFor` implementation.
I would just have preferred to name it like that - a guess - or at least slap a code comment on it if we can't fix it or don't need it to be reliable, so that we don't end up running in circles.